src/UglyToad.PdfPig/Writer/PdfMerger.cs

```diff
@@ -132,75 +130,59 @@ private class DocumentMerger
         {
             private const decimal DefaultVersion = 1.2m;
             
-            private readonly BuilderContext context = new BuilderContext();
-            private readonly List<IndirectReferenceToken> documentPages = new List<IndirectReferenceToken>();
-            private readonly IndirectReferenceToken rootPagesIndirectReference;
+            private readonly PdfStreamWriter context = new PdfStreamWriter();
+            private readonly List<IndirectReferenceToken> pagesTokenReferences = new List<IndirectReferenceToken>();
+            private readonly IndirectReferenceToken rootPagesReference;
 
             private decimal currentVersion = DefaultVersion;
-            private MemoryStream memory = new MemoryStream();
-
             private int pageCount = 0;
+            
+            private readonly Dictionary<IndirectReferenceToken, IndirectReferenceToken> referencesFromDocument =
```

Comment:
This dictionary would keep track of token that are being written from the current document being added to the resultant document. I put like so (globally, class scope), to avoid having to pass it as parameter, cluttering the parameters of `CopyToken`